### DISCUSSION TITLE
[A-L] (7 luglio 2023) Esercizio 2

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#76] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/11abc125-db48-4e97-af8a-e43664761613)

**Comment by Frisbon** [comment]
Ho paura di aver barato usando l'heap come istogramma, con lui abbiamo visto che si potevano usare per estrarre il min/max in $\log n$ perciò...
```python
def es2(A):
    H = heap()
    for element in A: #questo ciclo costa n*log(m) dove m<=n
        if H.value(element) == None:
            H.insert(element,1)
        else:
            H.update_value(element,H.value(element)+1)
    soglia = (2*len(A))//3
    if Heap.max_value() >= soglia:
        return "YES"
    else: return "NO"
```
Il costo finale (caso peggiore) è $O(n\cdot \log n)$

**Reply by Alby7503** [reply]
Fai bene ad avere paura 
![image](https://github.com/user-attachments/assets/04703012-b2d1-43b2-9a3b-4f83febb580f)

**Comment by cornflexxx** [comment]
Come suggerisce il costo computazionale $O(n \log n)$ il metodo per risolvere l'esercizio è usare un algoritmo divide et impera con equazione di ricorrenza: 
 $$T(n) = 2T(\frac{n}{ 2})+ \Theta ( n)$$
quindi dobbiamo dividere l'array in due sotto-array e ad ogni passo ricorsivo faremo un lavoro lineare.  Ad ogni passo della ricorsione dividiamo in due parti l'array fino ad arrivare ad array formati da un singolo elemento, questi elementi saranno sicuramente super majority element del loro sottoarray e quindi questo costituisce il caso base. 
Risalendo dalla ricorsione combiniamo le soluzioni delle  due metà volta per volta: sia $M_{sx}$ il $SME$  della metà di sinistra e $M_{dx}$ quello della metà destra, allora il $SME$ dell'array formato dall'unione delle due metà deve necessariamente essere $M_{sx}$ o $M_{dx}$ oppure non c'è nessun $SME$ in quel sotto array. Questo risultato deriva da un ragionamento molto semplice:
supponiamo ad un qualunque passo della ricorsione, che $m\in A_{sx}+A_{dx}$ non sia $SME$ del sotto array di sinistra $A_{sx}$ né tantomento lo sia per quello di destra $A_{dx}$, ne segue che in $A_{sx}$ e $A_{dx}$ sia presente in al più $\frac{n/2}{3}$ posizioni. Quindi in $A = A_{sx} +A_{dx}$ sarà presente in al massimo 
$$\frac{n/2}{3} + \frac{n/2}{3} = \frac{1}{3}n$$
 posizioni, ne segue che 
$$\frac{1}{3}n\le \frac {2}{3}n$$
e quindi non può essere un super majority element.

```
majority_element(A):
	if len(A) == 1:
		return A[0]
	m = len(A)/2
	M_dx = MajElt(A[m+1:])
	M_sx = MajElt(A[:m])
	dx, sx = 0, 0
        if M_dx and M_sx == Null : 
                return Null
	for i = 0,..,len(A)-1:
		if A[i]	== M_dx:
			dx += 1
		if A[i] == M_sx:
			sx += 1
	if sx > dx : Maj = M_sx
	else: Maj = M_dx
	
	if Maj >= ceil((len(A)/3)*2):
		return Maj
	else: 
		return Null
```

**Comment by daromacarlo** [comment]
```c
def majority(a : array){
	//definisco i casi base
	if len(a) == 3{
		if a[0] == a[1] or a[0] == a[2]{
			return a[0]
		}
		if a[1] == a[2]{
			return a[1]
		}
	}

	if len(a) == 0{
		return NULL
	}

	if len(a) == 1{
		return a[0]
	}
	
	if len(a) == 2{
		if a[0] == a[1]{
			return a[0]
		}
	}

	m = len(a) / 2
	L = majority(a[0:m])
	R = majority(a[m+1:len(a)-1])
	count = 0

	for i da 0 a len(a)-1{
		if a[i] == L{
			count ++
		}
	}
	if count == (2/3)*(len(a)){
		return L
	}

	count = 0

	for i da 0 a len(a)-1{
		if a[i] == R{
			count ++
		}
	}
	if count == (2/3)*(len(a)){
		return R
	}
	return NULL
}

def check(a)
	x = majority(a)
	if x != null{
		return "YES"
	}
	else return{
		false
	}

```
